data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/query/DefaultBindableParametersStoredQuery.java

```diff
@@ -158,10 +158,6 @@ public DataType getDataType() {
             // Empty collections / array should always set at least one value
             value = null;
             values = null;
-        } else if (values != null && values.size() == 1) {
```

Comment:
This is fix for this issue but also for #1788 we talked about earlier.

Reply:
Why did you include it in this PR?

Reply:
Because this is the fix for this issue as well.

Reply:
Without this, alternate fix is having this in AbstractHibernateOperations binding
```
if (Collection.class.isAssignableFrom(argumentType)) {
                        Collection<Object> paramCollection;
                        if (value == null) {
                            paramCollection = Collections.emptyList();
                        } else if (value instanceof Collection) {
                            paramCollection = (Collection<Object>) value;
                        } else {
                            // This is the case where collection with one element was passed and during binding collection moved to single value element
                            // but argumentType still remained collection
                            paramCollection = Collections.singletonList(value);
                        }
                        setParameterList(q, parameterName, paramCollection, argument.getFirstTypeVariable().orElse(Argument.OBJECT_ARGUMENT));
                    } else if (Object[].class.isAssignableFrom(argumentType)) {
                        Collection<Object> coll;
                        if (value == null) {
                            coll = Collections.emptyList();
                        } else if (value instanceof Collection) {
                            coll = (Collection<Object>) value;
                        } else if (value.getClass().isArray()) {
                            coll = Arrays.asList((Object[]) value);
                        } else {
                            // This is the case where array with one element was passed and during binding array values moved to single value element
                            // but argumentType still remained array
                            coll = Collections.singletonList(value);
                        }
                        setParameterList(q, parameterName, coll);
                    } else {
                        setParameter(q, parameterName, value, argument);
                    }
```

Reply:
I would expect the value to be recognized in `DefaultBindableParametersStoredQuery#expandValue`. What is wrong there?

Reply:
Expanding happens first and then it gets converted from Collection to the single element of the collection
```
List<Object> values = binding.isExpandable() ? expandValue(value, binding.getDataType()) : null;
        if (values != null && values.isEmpty()) {
            // Empty collections / array should always set at least one value
            value = null;
            values = null;
        } else if (values != null && values.size() == 1) {
            // Bind as one if the expanded collections is just one item
            value = values.iterator().next();
            values = null;
        }
```
which later confuses binder and tries to cast to collection where we have single element here in AbstractHibernateOperations
```
if (Collection.class.isAssignableFrom(argumentType)) {
                        setParameterList(q, parameterName, value == null ? Collections.emptyList() : (Collection<Object>) value, argument.getFirstTypeVariable().orElse(Argument.OBJECT_ARGUMENT));
                    }
```
because argumentType is still collection. No idea why would we convert from collection to single element. This change fixes this issue and one with custom Mongo find we talked before. The rest of the tests are passing so this seems fine to me.

Reply:
Maybe the correct would be to use `setParameterList` in the bind many

Reply:
Probably all those casts need to be corrected just to convert from collection

Reply:
The problem is when you change from collection to single element then bindOne is called 
```
if (values == null) {
            if (parameterConverter != null) {
                value = binder.convert(parameterConverter, value, argument);
            } else if (persistentProperty != null && !binding.isAutoPopulated()) {
                value = binder.convert(value, persistentProperty);
            }
            binder.bindOne(binding, value);
        }
```
and not sure why it is all needed in the first place? Just remove this collection change here and we are good.

Reply:
I guess that optimization brings way too many problems.